agent/consul/state/acl.go

```diff
@@ -560,6 +561,68 @@ func (s *Store) CanBootstrapACLToken() (bool, uint64, error) {
 	return false, out.(*IndexEntry).Value, nil
 }
 
+func (s *Store) resolveACLLinks(tx *memdb.Txn, links []agentpb.ACLLink, getName func(*memdb.Txn, string) (string, error)) (int, error) {
+	var numValid int
+	for linkIndex, link := range links {
+		if link.ID != "" {
+			name, err := getName(tx, link.ID)
+
+			if err != nil {
+				return 0, err
+			}
+
+			// the name doesn't matter here
+			if name != "" {
+				links[linkIndex].Name = name
+				numValid++
+			}
+		} else {
+			return 0, fmt.Errorf("Encountered an ACL resource linked by Name in the state store")
+		}
+	}
+	return numValid, nil
+}
+
+func (s *Store) fixupACLLinks(tx *memdb.Txn, original []agentpb.ACLLink, getName func(*memdb.Txn, string) (string, error)) ([]agentpb.ACLLink, bool, error) {
+	owned := false
+	links := original
+
+	cloneLinks := func(l []agentpb.ACLLink, copyNumLinks int) []agentpb.ACLLink {
+		clone := make([]agentpb.ACLLink, copyNumLinks)
+		copy(clone, l[:copyNumLinks])
+		return clone
+	}
+
+	for linkIndex, link := range original {
+		name, err := getName(tx, link.ID)
```

Comment:
Did you want to carry over this guard?
```
		if link.ID == "" {
			return nil, fmt.Errorf("Detected corrupted <something> within the state store - missing link ID")
		}
```

Reply:
That is now the responsibility of the `getName` function passed in. If they care then they will return an error.